// react-formic/lib/components/controls/SubmitButton.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _connectForm = require('../connectors/connectForm');

var _connectForm2 = _interopRequireDefault(_connectForm);

var _validationUtils = require('../../utils/validationUtils');

var _validity = require('../../constants/validity');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SubmitButton = function SubmitButton(props) {
  var className = props.className;
  var children = props.children;
  var form = props.form;
  var _onClick = props.onClick;
  var Tag = props.tag;

  var formValidity = (0, _validationUtils.getFormValidity)(form);
  var isValid = formValidity === _validity.VALID;

  var proxyProps = _extends({}, props, {
    className: (0, _classnames2.default)(className, {
      invalid: formValidity === _validity.INVALID,
      pending: formValidity === _validity.PENDING,
      valid: isValid
    }),
    disabled: !isValid,
    onClick: function onClick(event) {
      event.preventDefault();
      if (isValid) {
        _onClick(event);
      }
    }
  });

  return _react2.default.createElement(
    Tag,
    proxyProps,
    children
  );
};

SubmitButton.propTypes = {
  children: _react2.default.PropTypes.node,
  className: _react2.default.PropTypes.string,
  form: _react2.default.PropTypes.object,
  tag: _react2.default.PropTypes.string,
  onClick: _react2.default.PropTypes.func
};

SubmitButton.defaultProps = {
  tag: 'a'
};

exports.default = (0, _connectForm2.default)(function (form) {
  return { form: form };
})(SubmitButton);
